src/shogun/base/SGObject.h

```diff
@@ -1022,5 +1015,161 @@ class CSGObject
 };
+
+template <class T>
+class ObservedValueTemplated;
+
+/**
+ * Observed value which is emitted by algorithms.
+ */
+class ObservedValue : public CSGObject
+{
+public:
+	/**
+	 * Constructor
+	 * @param step step
+	 * @param name name of the observed value
+	 */
+	ObservedValue(int64_t step, std::string name);
+
+	/**
+	 * Destructor
+	 */
+	~ObservedValue(){};
+
+#ifndef SWIG
+	/**
+	* Return a any version of the stored type.
+	* @return the any value.
+	*/
+	virtual Any& get_any()
+	{
+		return m_any_value;
+	}
+#endif
+
+	/** @return object name */
+	virtual const char* get_name() const
+	{
+		return "ObservedValue";
+	}
+
+protected:
+	/** ObservedValue step (used by Tensorboard to print graphs) */
+	int64_t m_step;
+	/** Parameter's name */
+	std::string m_name;
+	/** Untyped value */
+	Any m_any_value;
+};
+
+/**
+ * Templated specialisation of ObservedValue that stores the actual data.
+ * @tparam T the type of the observed value
+ */
+template <class T>
+class ObservedValueTemplated : public ObservedValue
+{
+
+public:
+	/**
+	 * Constructor
+	 * @param step step
+	 * @param name the observed value's name
+	 * @param value the observed value
+	 */
+	ObservedValueTemplated(
+		int64_t step, std::string name, std::string description, T value)
+		: ObservedValue(step, name), m_observed_value(value)
+	{
+		this->watch_param(
+			name, &m_observed_value,
+			AnyParameterProperties(description, ParameterProperties::NONE));
+		m_any_value = make_any(m_observed_value);
+	}
+
+	/**
+	 * Constructor which takes AnyParameterProperties for the observed value
+	 * @param step step
+	 * @param name the observed value's name
+	 * @param value the observed value
+	 * @param properties properties of that observed value
+	 */
+	ObservedValueTemplated(
+		int64_t step, std::string name, T value,
+		AnyParameterProperties properties)
+		: ObservedValue(step, name), m_observed_value(value)
+	{
+		this->watch_param(name, &m_observed_value, properties);
+		m_any_value = make_any(m_observed_value);
+	}
+
+	/**
+	 * Destructor
+	 */
+	~ObservedValueTemplated(){};
+
+private:
+	/**
+	 * Templated observed value
+	 */
+	T m_observed_value;
+};
+
+template <typename T,
+	      typename std::enable_if_t<!is_string<T>::value>* = nullptr>
+void CSGObject::put(const Tag<T>& _tag, const T& value, int64_t step) noexcept(
+	false)
+{
+	if (has_parameter(_tag))
+	{
+		auto parameter_value = get_parameter(_tag).get_value();
+		if (!parameter_value.cloneable())
+		{
+			SG_ERROR(
+				"Cannot put parameter %s::%s.\n", get_name(),
+				_tag.name().c_str());
+		}
+		try
+		{
+			any_cast<T>(parameter_value);
+		}
+		catch (const TypeMismatchException& exc)
+		{
+			SG_ERROR(
+				"Cannot put parameter %s::%s of type %s, incompatible "
+				"provided type %s.\n",
+				get_name(), _tag.name().c_str(), exc.actual().c_str(),
+				exc.expected().c_str());
+		}
+		ref_value(value);
+		update_parameter(_tag, make_any(value));
+
+		observe_tag<T>(step, _tag.name());
```

Comment:
need to find another solution with the `step`, it cannot be part of the put API

Reply:
Maybe we do this
```
int step = -1;
Tag<int> tag("current_iteration") // basically refers to iterative machine which is what we wanna use for algorithms with iteration counters
if (has(tag))
  step = get(tag);
```
A hacky solution. This could be more elaborate, for example machines with iteration counters could register methods/lambdas with the base class upon construction to share the iteration.
@gf712 @lisitsyn thoughts?

Reply:
What about adding a singleton counter class? The only thing it will do is provide an incremental number every time it is called. A sample usage could be like:
```c++
int64_t step = IncrementalCounter::get_instance().next_step();
observe_tag<T>(step, _tag.name())
```
Therefore, all the emitted values will be somehow syncronized.

Reply:
thread safety would be a concern with singletons. The auto counter raises issues when more than one thing are observed. 

The problem with my above idea is runtime costs. We could later do some benchmarking how much overhead the `has`/`get` solution above causes. If it is significant, I would suggest to add a step protected member variable in `CMachine`. But let's do that in another PR so we can move on here, just do the `get` thingi.

```
SG_FORCED_INLINE int get_step()
{
int step = -1;
Tag<int> tag("current_iteration") // could be changed later to local field
if (has(tag))
  step = get(tag);
return step;
}
```
and then use that, which gives us options to overload/virtual it later on, or use a local variable

Reply:
Updated with `get_step()`.